src/QirRuntime/public/QuantumApiBase.hpp

```diff
@@ -0,0 +1,263 @@
+#pragma once
+
+#include <exception>
+
+#include "IQuantumApi.hpp"
+
+namespace Microsoft
+{
+namespace Quantum
+{
+    class CQuantumApiBase : public IQuantumApi

```

Comment:
This class promotes building an backend for which certain methods permanently throw a runtime exception with the expectation that these methods are simply not used. We should not promote this, since this is rather fragile design. I would like to remove this.

Reply:
This class allows quick experimentation with building a new simulator and also makes test code _much more_ readable. I can rename it to CQuantumApiNoneImplemented to make it more clear that the class isn't recommended to be used as a permanent base class in production code and replace it in the toffoli/fullstate simulators with direct inheritance from the interface and explicit "no supported" rather than "not implemented" throws on the relevant methods.

Reply:
The class is used by production tests and experimental features and their tests so it would create a wrong dependency direction to move it into either. While some tools do support refactoring of C++ code, as far as I know there is no widely accepted one yet. The class allows to start on a new simulator quickly, being able to compile and test it with minimal effort. Turning the code into production is also simple by replacing the inheritance from this class to interface and implementing everything that compiler would complain about.

Reply:
Before we can merge this into main, I think we need to replace the interface as well as this class. Throwing "not supported" permanently instead of "not implemented" doesn't solve the problem, which is that the base interface currently pre-selects a specific quantum instruction set where the quantum instruction set should be completely open and unspecified,

A code generation solution for  like we're enabling for managed simulators is the right long-term solution, allowing every target to specify the quantum instruction set (and common gate decompositions) it supports in a standard way.

Reply:
While I agree with the idea that this interface and related base classes will be heavily refactored to take advantage of the redesign of target packages and the introduction of the code generation components, neither of those parts are done yet. I don't think we should force this PR, which introduces code that is not yet part of the production workflow, to wait on all the other related work to be finished before it gets merged. I would be more in favor of merging this now in its working form, where it mimics the current patterns in the production infrastructure, and then iterating on it in main as the other work comes in. That way we don't continue developing in the mode where large chunks of code live in feature branches waiting for integration with other large chunks also in feature branches.

Reply:
I believe this thread should be resolved with the plan we defined yesterday. To summarize: We will merge this PR before any of the automation infrastructure exists, and we will roughly align the structure of the code to how the code is expected to be structured in the the future when parts of it are autogenerated instead of manually implemented. This PR contains the manual implementation and execution support for one specific target package, and all further target packages will be supported once we have more automation built out. Let me know if there are still open questions.

Reply:
Split IQuantumApi into ISimulator (qubit and result management) and IQuantumGateSet both of which are currently implemented by the full state simulator. When the automatic generation of target gate sets is available, the latter of the two interfaces will be deprecated.

CQuantumApiBase is now named SimulatorStub and not used in production code.